gadget/update.go

```diff
@@ -326,6 +327,95 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
+// the specified volume can be laid out exactly to the disk device, returning an
+// error if the two are not compatible.
+func DiskVolumeDeviceTraitsForDevice(laidOutVolume *LaidOutVolume, dev string) (res DiskVolumeDeviceTraits, err error) {
+	vol := laidOutVolume.Volume
+
+	// get the disk layout for this device
+	diskLayout, err := OnDiskVolumeFromDevice(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot read %v partitions for candidate volume %s: %v", dev, vol.Name, err)
+	}
+
+	// ensure that the on disk and the laid out version are actually compatible
+	opts := &EnsureLayoutCompatibilityOptions{
+		AssumeCreatablePartitionsCreated: true,
+	}
+	if err := EnsureLayoutCompatibility(laidOutVolume, diskLayout, opts); err != nil {
+		return res, fmt.Errorf("volume %s is not compatible with disk %s: %v", vol.Name, dev, err)
+	}
+
+	// also get a Disk{} interface for this device
+	disk, err := disks.DiskFromDeviceName(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot get disk for device %s: %v", dev, err)
+	}
+
+	mappedStructures := make([]DiskStructureDeviceTraits, 0, len(diskLayout.Structure))
+
+	diskPartitions, err := disk.Partitions()
+	if err != nil {
+		return res, fmt.Errorf("cannot get partitions for disk device %s: %v", dev, err)
+	}
+
+	// make an easy map of device nodes to partitions for lookup
+	diskPartitionsByDeviceNode := map[string]disks.Partition{}
+	for _, p := range diskPartitions {
+		diskPartitionsByDeviceNode[p.KernelDeviceNode] = p
+	}
+
+	for _, structure := range diskLayout.Structure {
+		part, ok := diskPartitionsByDeviceNode[structure.Node]
+		if !ok {
+			// unexpected error - somehow this structure in the OnDiskVolume
+			// does not exist in the same physical disk we built for it using
+			// udev and disks.DiskFromDeviceName (which internally uses the same
+			// code as OnDiskVolumeFromDevice).
+			return res, fmt.Errorf("internal error: inconsistent disk structures from OnDiskVolume and disks.Disk: partition for device node %s missing", structure.Node)
+		}
+		ms := DiskStructureDeviceTraits{
+			Size:               quantity.Size(part.SizeInBytes),
+			Offset:             quantity.Offset(part.StartInBytes),
+			PartitionUUID:      part.PartitionUUID,
+			OriginalKernelPath: part.KernelDeviceNode,
+			OriginalDevicePath: part.KernelDevicePath,
+			PartitionType:      part.PartitionType,
+			PartitionLabel:     part.PartitionLabel,  // this will be empty on dos disks
+			FilesystemLabel:    part.FilesystemLabel, // blkid encoded
+			FilesystemUUID:     part.FilesystemUUID,  // blkid encoded
+			FilesystemType:     part.FilesystemType,
+		}
+
+		mappedStructures = append(mappedStructures, ms)
+
+		// delete this partition from the map
+		delete(diskPartitionsByDeviceNode, structure.Node)
+	}
+
+	// check that we don't have any extra partitions on the disk that are not
+	// accounted for in the layout
+	if len(mappedStructures) != len(diskPartitions) {
```

Comment:
both diskPartitions and mappedStructures come from the actual disk? what is the situation this is trying to consider?

Reply:
ah you're right this is wrong, it was meant to check against `laidOutVolume` not `diskLayout` but the idea behind this check is that the size of mappedStructures is supposed to come from the `laidOutVolume` which is not from the disk and then check that against what comes from the disk verbatim via udev in `disk.Partitions()`